_posts/2017-01-23-cpp-strings.md

```diff
@@ -0,0 +1,150 @@
+---
+title: Modeling strings with libpmemobj C++ bindings
+author: RobDickinson
+layout: post
+---
+
+C++ developers using libpmemobj have multiple options for modeling strings,
+depending on the size of the strings and whether they are fixed or varying
+in length.
+
+For small fixed-length strings, use of `p<char[size]>` is efficient and easy.
+For large or variable-length strings, or for structs or arrays of such
+strings, then things get a little more complicated.
+
+#### Avoiding use of std::string
+
+For starters, never use `p<std::string>` or `persistent_ptr<std::string>`.
+These forms might seem a logical place to start, but these implementations
+will not be power-fail safe (at best) and might be unstable (at worst).
+This advice applies to using `p<>` or `persistent_ptr<>` wrappers around
+any complex types that perform their own memory management.
+Don't make this obvious mistake!
+
+#### Avoiding arrays of persistent pointers to single chars
+
+Don't model a variable-length string with `p<char> X[]`. This is power-fail
+safe, but wastes a lot of memory. Let's consider what's happening behind the
+scenes here. With this representation, each char in the variable-length
+string will be stored as one char in persistent memory, plus another 16 bytes
+for the persistent pointer to that char. So writing a single char results in
+16x more bytes written than necessary. Reading a single char results in 16x
+more bytes read than necessary, since reading every persistent char requires
+dereferencing a 16-byte persistent pointer. Obviously this is quite wasteful.
+
+#### Using fixed-length strings inside persistent structs
+
+While `p<char[size]>` is the obvious starting point for modeling fixed-size
+strings, we'll frequently be modeling strings within nested persistent
+structs and classes as well. So we'll use a `p<>` wrapper around a complex
+type that contains one or more strings, as with `MyAssocations` below.
+
+```
+class MyAssociation {
+public:
+  void set_key(std::string* key);
+  void set_value(std::string* value);
+private:
+  char key[SIZE];
+  char value[SIZE];
+};
+
+struct MyAssociations {
+  p<MyAssociation> preferred_association;
+  p<MyAssociation> alternate_associations[SIZE];
+};
+```
+
+The downside of this approach is that the `set_key` and `set_value` methods
+must call `pmemobj_tx_add_range_direct` prior to modifying their respective
+internal fields.
+
+However, calling `pmemobj_tx_add_range_direct` is not required
+when modifying a newly allocated object, as the entire memory range of a new
+object is already in the current transaction.
+
+So for best performance, we could add `init_key` and `init_value` setter
+methods, which don't call `pmemobj_tx_add_range_direct` by convention, and so
+would only be safe to use in the context of a newly allocated object. These
+would be slightly faster than `set_key` and `set_value`.
```

Comment:
We have discussed this at length today with @pbalcer and @marcinslusarz, and although this is perfectly valid code, adding a new API for this slight improvement is probably not the best idea. It will complicate the client code with little performance gain.